text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+
+## async/await
+
+This crate is by nature very much dependent on events happening outside of the control of the program, e.g. listening for incoming connections from peers, waiting for packets on a specific socket, etc. Hence, this crate under the hood makes heavy use of Rust's concurrency abstractions. Luckily, with the stabilization of the `async/await` syntax, writing asynchronous code has become almost as easy to read, write, and maintain as synchronous code. In an experimental implementation of this crate the [`async_std`](https://github.com/async-rs/async-std) library was used, which comes with asynchronous drop-in replacements for their synchronous equivalents found in Rust's standard library `std`. Additionally, asynchronous mpsc channels were taken from the [`futures`](https://github.com/rust-lang/futures-rs) crate.
+
+## Message passing
+
+This crate favors message passing via channels over globally shared state and locks. Instead of keeping the list of endpoints in a globally accessible hashmap this crate separates and moves such state into workers that run asynchronously, and are listening for commands and events to act upon, and also notify listeners by sending events.
+
+## Initialization
+
+In order for this crate to be used, it has to be initialized first by providing a `NetworkConfig`:
+
+```rust
+struct NetworkConfig {
+    // The server or binding TCP port that remote endpoints can connect and send to.
+    binding_port: u16,
+    // The binding address that remote endpoints can connect and send to.
+    binding_addr: IpAddr,
+    // The interval between two connections attempts in case of connection losses.
+    reconnect_interval: u64,
+    // The maximum message length in terms of bytes.
+    max_message_length: usize
+
+    ...
+}
+```
+
+This has the advantage of being easily extensible and keeping the signature of the static `init` function small:
+
+```rust
+fn init(config: NetworkConfig) -> (Network, Events, Shutdown) { ... }
+```
+
+This function returns a tuple struct that allows the consumer of this crate to send commands (`Network`), listen to events (`Events`), and gracefully shutdown all running asynchronous tasks that were spawned within this system (`Shutdown`).
+
+## `Port`, `Address`, `Protocol`, and `Url`
+
+In order to send a message to an endpoint, a node needs to know the endpoint's address and the communication protocol it uses. This crate provides the following abstractions to deal with that:
+
+* `Port`: a 0-cost wrapper around a `u16`, which is introduced for type safety and better readability:
+
+    ```rust
+    struct Port(pub u16);
+    ```
+
+    For convenience, `Port` dereferences to `u16`.
+
+* `Address`: a 0-cost wrapper around a `SocketAddr` which provides an adjusted, but overall simpler API than its inner type:
+
+    ```rust
+    #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)
+    struct Address(SocketAddr);
+    ```
```

Comment:
For some reason, this snippet is not syntactically colored 🤔

Reply:
haha, the closing bracket `]` is missing. Thanks for pointing this out. :smile:

Reply:
Ah right, didn't even see it

Reply:
Solved.